### PR TITLE
Reworking of ActionResult class's role and structure

### PR DESCRIPTION
**Pull request by antonovchtd**
Why 39 files?
Changes to API (return parameter) were introduced.
Before, there were two ActionResult classes, one which was used by *both* model and Actions (commands), and another (based on protobuf), for the transport layer. Now, Core::ModelInquiryResult is a new simplified class that is used by the model and transport. It's now on protobuf and uses oneof of status and ID, instead of both at the same time.

Actions still use a class that's called ActionResult, but it now contains Core::ModelInquiryResult as an optional parameter (union of Core::ModelInquiryResult, Core::TaskEntity, std::vector<Core::TaskEntity>). 

Also, Task.proto renamed to Core.proto.

### CODE REVIEW

File: src/model/TaskManager.cpp [hunk 1]
```diff
@@ -110,101 +120,140 @@ std::vector<Core::TaskEntity> TaskManager::getTaskWithSubtasks(const Core::TaskI
...
     }
-    for (auto const &ch : ChildrenOf(id))
-        Uncomplete(ch);
-    return {ActionResult::Status::SUCCESS, id};
+
+    return result;
 }
 
-ActionResult TaskManager::IsPresent(const Core::TaskID &id) const{
+Core::ModelInquiryResult TaskManager::IsPresent(const Core::TaskID &id) const{
```
Comment (KharchenkoMaks) [inline src/model/TaskManager.cpp]: Methods with prefix `Is` should return `bool`.

File: src/model/TaskManager.cpp [hunk 2]
```diff
@@ -46,25 +46,35 @@ std::optional<Core::TaskID> TaskManager::ParentOf(const Core::TaskID &id) const
...
-    if (IsPresent(id))
-        return {ActionResult::Status::DUPLICATE_ID, id};
+Core::ModelInquiryResult TaskManager::Add(const Core::Task &t) {
+    Core::ModelInquiryResult result;
+    auto id = gen_->genID();
 
-    tasks_.insert(std::make_pair(id, std::make_pair(t, Node())));
-    return {ActionResult::Status::SUCCESS, id};
+    if (ToBool(IsPresent(id))) {
+        result.set_status(Core::ModelInquiryResult_Status_DUPLICATE_ID);
```
Comment (KharchenkoMaks) [inline src/model/TaskManager.cpp]: Case when `IdGenerator` has created duplicate id is internal error of `TaskManager`, why should user of `TaskManager` API know about it? He didn't enter any id's that could be duplicate.

File: src/ui/actions/ActionResult.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+ActionResult::ActionResult(const std::vector<Core::TaskEntity> &vec) :
+              tasks{vec}, type_id{kVector} {
+}
+
+ActionResult::~ActionResult() {
+    switch (type_id) {
+        case kResult:
+            (&model_result)->Core::ModelInquiryResult::~ModelInquiryResult();
```
Comment (KharchenkoMaks) [inline src/ui/actions/ActionResult.cpp]: Maybe not the best solution.

File: src/ui/actions/LoadFromFileAction.cpp [hunk 4]
```diff
@@ -10,14 +10,22 @@ LoadFromFileAction::LoadFromFileAction(const std::string &filename) : filename_{
...
-        return {ActionResult::Status::TAKES_ARG, std::nullopt};
-    persister_ = std::unique_ptr<Persister>(new FilePersistence{filename_});
+    Core::ModelInquiryResult result;
+
+    if (filename_.empty()) {
+        result.set_status(Core::ModelInquiryResult_Status_TAKES_ARG);
+        return result;
+    }
+
+    persister_ = std::unique_ptr<Persister>(std::make_unique<FilePersistence>(filename_));
```
Comment (KharchenkoMaks) [inline src/ui/actions/LoadFromFileAction.cpp]: Just `persister_ = std::make_unique<FilePersistence>(filename_)`???

**Review by KharchenkoMaks — CHANGES_REQUESTED** [review]

File: src/model/TaskManager.cpp (hunk 1 above)
Comment (antonovchtd) [reply src/model/TaskManager.cpp]: I will change the output parameter in a separate PR

File: src/model/TaskManager.cpp (hunk 2 above)
Comment (antonovchtd) [reply src/model/TaskManager.cpp]: Fair point, I removed this check form Add and AddSubtask (also tests)

File: src/ui/actions/ActionResult.cpp (hunk 3 above)
Comment (antonovchtd) [reply src/ui/actions/ActionResult.cpp]: maybe, but union implicitly deletes the default destructor here

File: src/ui/actions/LoadFromFileAction.cpp (hunk 4 above)
Comment (antonovchtd) [reply src/ui/actions/LoadFromFileAction.cpp]: kk

File: proto/Core.proto [hunk 5]
```diff
@@ -30,4 +30,25 @@ message TaskEntity {
 
 message Label {
   string label = 1;
+}
+
+message ModelInquiryResult {
```
Comment (vladimir-moiseiev) [inline proto/Core.proto]: `Inquiry` is more a human term, which is hard to pronounce.
consider `Request`

File: src/ui/actions/ActionResult.cpp (hunk 3 above)
Comment (vladimir-moiseiev) [reply src/ui/actions/ActionResult.cpp]: Already comment in other review: Never call destructors directly!!!!!!!!

**Review by vladimir-moiseiev — CHANGES_REQUESTED** [review]

File: src/ui/actions/ActionResult.cpp (hunk 3 above)
Comment (antonovchtd) [reply src/ui/actions/ActionResult.cpp]: Fixed - refused using union altogether, employed std::optional instead

File: proto/Core.proto (hunk 5 above)
Comment (antonovchtd) [reply proto/Core.proto]: renamed

**Pull request merged**